File: frontend/src/services/customers.ts
import { api } from "@/services/api"
import {
  DEFAULT_PAGE_SIZE,
  mapPaginated,
  type ListPaginationParams,
  type PaginatedApi,
  type PaginatedList,
} from "@/types/pagination"
import type {
  Customer,
  CustomerCreatePayload,
  CustomerDetail,
  CustomerUpdatePayload,
} from "@/types/customers"

export type CustomerListParams = ListPaginationParams & {
  query?: string
}

export async function listCustomers(
  params: CustomerListParams = {},
): Promise<PaginatedList<Customer>> {
  const trimmed = params.query?.trim()
  const { data } = await api.get<PaginatedApi<Customer>>("/customers", {
    params: {
      q: trimmed || undefined,
      page: params.page ?? 1,
      page_size: params.pageSize ?? DEFAULT_PAGE_SIZE,
    },
  })
  return mapPaginated(data)
}

export async function getCustomer(customerId: string): Promise<CustomerDetail> {
  const { data } = await api.get<CustomerDetail>(`/customers/${customerId}`)
  return data
}

export async function createCustomer(
  payload: CustomerCreatePayload,
): Promise<Customer> {
  const { data } = await api.post<Customer>("/customers", payload)
  return data
}

export async function updateCustomer(
  customerId: string,
  payload: CustomerUpdatePayload,
): Promise<Customer> {
  const { data } = await api.patch<Customer>(
    `/customers/${customerId}`,
    payload,
  )
  return data
}

export async function deleteCustomer(customerId: string): Promise<void> {
  await api.delete(`/customers/${customerId}`)
}

File: backend/app/api/v1/endpoints/customers.py
"""Customer directory API endpoints."""

from typing import Annotated

from fastapi import APIRouter, Depends, Query, Response, status

from app.api.deps import DEFAULT_LIST_PAGE, DEFAULT_LIST_PAGE_SIZE, PageParam, PageSizeParam
from app.auth.deps import get_current_user, get_current_workspace
from app.schemas.auth import UserResponse
from app.schemas.customer import (
    CustomerCreateRequest,
    CustomerDetailResponse,
    CustomerResponse,
    CustomerUpdateRequest,
)
from app.schemas.pagination import PaginatedResponse
from app.services import customer_service
from app.services.workspace_service import WorkspaceContext

router = APIRouter(prefix="/customers", tags=["customers"])


@router.get("", response_model=PaginatedResponse[CustomerResponse])
async def list_customers(
    current_workspace: Annotated[WorkspaceContext, Depends(get_current_workspace)],
    q: Annotated[
        str | None,
        Query(max_length=200, description="Search name, email, or company"),
    ] = None,
    page: PageParam = DEFAULT_LIST_PAGE,
    page_size: PageSizeParam = DEFAULT_LIST_PAGE_SIZE,
) -> PaginatedResponse[CustomerResponse]:
    return await customer_service.list_customers(
        current_workspace.id,
        query=q,
        page=page,
        page_size=page_size,
    )


@router.post(
    "",
    response_model=CustomerResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_customer(
    payload: CustomerCreateRequest,
    current_user: Annotated[UserResponse, Depends(get_current_user)],
    current_workspace: Annotated[WorkspaceContext, Depends(get_current_workspace)],
) -> CustomerResponse:
    return await customer_service.create_customer(
        workspace_id=current_workspace.id,
        owner_id=current_user.id,
        payload=payload,
    )


@router.get("/{customer_id}", response_model=CustomerDetailResponse)
async def get_customer(
    customer_id: str,
    current_workspace: Annotated[WorkspaceContext, Depends(get_current_workspace)],
) -> CustomerDetailResponse:
    return await customer_service.get_customer(
        customer_id,
        current_workspace.id,
    )


@router.patch("/{customer_id}", response_model=CustomerResponse)
async def update_customer(
    customer_id: str,
    payload: CustomerUpdateRequest,
    current_workspace: Annotated[WorkspaceContext, Depends(get_current_workspace)],
) -> CustomerResponse:
    return await customer_service.update_customer(
        customer_id,
        current_workspace.id,
        payload,
    )


@router.delete(
    "/{customer_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_class=Response,
)
async def delete_customer(
    customer_id: str,
    current_workspace: Annotated[WorkspaceContext, Depends(get_current_workspace)],
) -> Response:
    await customer_service.delete_customer(customer_id, current_workspace.id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)
